### PR TITLE
Improve loom features support

### PR DESCRIPTION
**Pull request by PerfectSlayer**
# What Does This Do

This PR brings support for the Java structured concurrency API (currently in preview) and refactors the java concurrent instrumentations:

* Introduce a new Java module for JDK21 `:java-concurrent-21`,
  * Introduce a dedicated test suite for preview APIs: `previewTest`,
* Move instrumentation classes per features into dedicated packages,
* Improve virtual threads tests to no more use `latestDep` and be part of the `check` Gradle goal,
* Fix test cases where thread pool is `null`.
* Update instrumentation documentation about helper classes for bootstrap instrumentation.

# Motivation

The goal is to improve our support for (current and incoming) Loom features.

# Additional Notes

Jira ticket: [PROJ-IDENT]

### CODE REVIEW

File: dd-java-agent/instrumentation/java-concurrent/java-concurrent-21/src/main/java/datadog/trace/instrumentation/java/concurrent/structuredconcurrency/StructuredTaskScopeInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import net.bytebuddy.asm.Advice;
+
+/**
+ * This instrumentation captures the active span scope at StructuredTaskScope task creation
+ * (SubtaskImpl). The scope is then activate and close through the Runnable instrumentation
+ * (SubtaskImpl implementation Runnable).
+ */
+@SuppressWarnings("unused")
+@AutoService(InstrumenterModule.class)
+public class StructuredTaskScopeInstrumentation extends InstrumenterModule.Tracing
```
Comment (amarziali) [inline dd-java-agent/instrumentation/java-concurrent/java-concurrent-21/src/main/java/datadog/trace/instrumentation/java/concurrent/structuredconcurrency/StructuredTaskScopeInstrumentation.java:28]: You can shortcut the matching by implementing  extending `isEnabled` and checking if the jdk is at least 21

```
  @Override
  public boolean isEnabled() {
    return Platform.isJavaVersionAtLeast(21)
        && super.isEnabled() ;
```

File: dd-java-agent/instrumentation/java-concurrent/java-concurrent-21/build.gradle [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+tasks.named("check").configure {
+  dependsOn "previewTest"
+}
+
+dependencies {
+  testImplementation project(':dd-java-agent:instrumentation:trace-annotation')
+}
+
+project.tasks.withType(AbstractCompile).configureEach {
+  setJavaVersion(it, 21)
```
Comment (amarziali) [inline dd-java-agent/instrumentation/java-concurrent/java-concurrent-21/build.gradle:49]: I'm not sure if it's useful here but should we set code compatibility to 8 for the advice generated bytecode? 
e.g.
```
sourceCompatibility = JavaVersion.VERSION_1_8
  targetCompatibility = JavaVersion.VERSION_1_8
```

File: dd-java-agent/instrumentation/java-concurrent/build.gradle [hunk 3]
```diff
@@ -10,17 +6,11 @@ muzzle {
 
 apply from: "$rootDir/gradle/java.gradle"
 
-addTestSuite('latestDepTest')
```
Comment (amarziali) [inline dd-java-agent/instrumentation/java-concurrent/build.gradle:13]: The latestDep contained a test around `Exectutors.newVirtualThreadPerTaskExecutor` (only available on 21+). Why this test suite has been removed?

Comment (PerfectSlayer) [reply dd-java-agent/instrumentation/java-concurrent/build.gradle:13]: It was moved into the default `test` suite of `:dd-java-agent:instrumentation:java-concurrent:java-concurrent-21` as it applies only to Java 21+

File: dd-java-agent/instrumentation/java-concurrent/java-concurrent-21/src/main/java/datadog/trace/instrumentation/java/concurrent/structuredconcurrency/StructuredTaskScopeInstrumentation.java (hunk 1 above)
Comment (PerfectSlayer) [reply dd-java-agent/instrumentation/java-concurrent/java-concurrent-21/src/main/java/datadog/trace/instrumentation/java/concurrent/structuredconcurrency/StructuredTaskScopeInstrumentation.java:28]: Should I do it also for the existing `datadog.trace.instrumentation.java.concurrent.virtualthread.TaskRunnerInstrumentation` instrumentation? (dedicated virtual threads)

File: dd-java-agent/instrumentation/java-concurrent/build.gradle (hunk 3 above)
Comment (amarziali) [reply dd-java-agent/instrumentation/java-concurrent/build.gradle:13]: Oh nice I just skipped this reading the diff thanks

File: dd-java-agent/instrumentation/java-concurrent/java-concurrent-21/build.gradle (hunk 2 above)
Comment (amarziali) [reply dd-java-agent/instrumentation/java-concurrent/java-concurrent-21/build.gradle:49]: that should be set only on the advice compilation so AbstractCompile is too wide. I would narrowed the different tasks as done in other advices. Again i'm not sure that we need this compatibility but I thought that we assume injecting 1.8 compatible bytecode

Comment (PerfectSlayer) [reply dd-java-agent/instrumentation/java-concurrent/java-concurrent-21/build.gradle:50]: I improved the gradle configuration to limit the 1.8 compatibility to the instrumentation code compilation.

**Review by amarziali — APPROVED** [review]
lgtm

**Pull request merged**